### PR TITLE
Support for `MOD`, `PMOD` and `PYMOD` for `decimal32/64/128`

### PR DESCRIPTION
**Pull request by codereport**
Resolves `libcudf` `MOD` and `PMOD` portion of https://github.com/rapidsai/cudf/issues/7132

To Do:
* [x] `MOD`
* [x] `PMOD`
* [x] `PYMOD`
* [x] Unit tests
* [x] Different scaled tests

### CODE REVIEW

**Review by hyperbolic2346 — COMMENTED** [review]
Copyrights need updating here.

File: cpp/tests/binaryop/binop-compiled-fixed_point-test.cpp [hunk 1]
```diff
@@ -33,14 +33,14 @@
 namespace cudf::test::binop {
 
 template <typename T>
-struct FixedPointCompiledTestBothReps : public cudf::test::BaseFixture {
+struct FixedPointCompiledTestAllReps : public cudf::test::BaseFixture {
```
Comment (robertmaynard) [inline cpp/tests/binaryop/binop-compiled-fixed_point-test.cpp:36]: I think we can just go down to ` FixedPointCompiledTest` since we see the test suite is using `cudf::test::FixedPointTypes`

Comment (codereport) [reply cpp/tests/binaryop/binop-compiled-fixed_point-test.cpp:36]: Yea that is a good point. Also prevents churn if we ever get 256bit decimal types :)

File: cpp/src/binaryop/binaryop.cpp [hunk 2]
```diff
@@ -110,7 +110,9 @@ bool is_comparison_binop(binary_operator op)
  */
 bool is_supported_fixed_point_binop(binary_operator op)
 {
-  return is_basic_arithmetic_binop(op) or is_comparison_binop(op);
+  return is_basic_arithmetic_binop(op) or is_comparison_binop(op) or  //
+         op == binary_operator::MOD or                                //
+         op == binary_operator::PMOD;
```
Comment (bdice) [inline cpp/src/binaryop/binaryop.cpp:115]: Is there a reason we didn't add PYMOD? It's closely related to MOD/PMOD and would be the expected behavior for Python users with fixed-point columns.

File: cpp/tests/binaryop/binop-compiled-fixed_point-test.cpp [hunk 3]
```diff
@@ -684,6 +684,122 @@ TYPED_TEST(FixedPointCompiledTestBothReps, FixedPointBinaryOpThrows)
                cudf::logic_error);
 }
 
+TYPED_TEST(FixedPointCompiledTestAllReps, FixedPointBinaryOpModSimple)
```
Comment (bdice) [inline cpp/tests/binaryop/binop-compiled-fixed_point-test.cpp:687]: Are there tests for the case where the scales are not equal? If not, we need them.

**Review by bdice — CHANGES_REQUESTED** [review]
Comments attached.

File: cpp/include/cudf/fixed_point/fixed_point.hpp [hunk 4]
```diff
@@ -440,6 +440,21 @@ class fixed_point {
...
+   *
+   * If `_scale`s are equal, `_value`s are modulo-ed <br>
+   * If `_scale`s are not equal, number with smaller `_scale` is shifted to the
+   * greater `_scale`, and then `_value`s are modulo-ed
+   *
+   * @tparam Rep1 Representation type of number being modulo-ed to `this`
+   * @tparam Rad1 Radix (base) type of number being modulo-ed to `this`
+   * @return The resulting `fixed_point` number
+   */
+  template <typename Rep1, Radix Rad1>
```
[the remark this answers is not in the document]
Comment (codereport) [reply cpp/include/cudf/fixed_point/fixed_point.hpp:454]: I believe this is to avoid conflict with the class templates on `fixed_point`

File: cpp/src/binaryop/compiled/operation.cuh [hunk 5]
```diff
@@ -162,6 +162,16 @@ struct PMod {
     if (rem < 0) rem = std::fmod(rem + yconv, yconv);
     return rem;
   }
+
+  template <typename TypeLhs,
+            typename TypeRhs,
+            std::enable_if_t<cudf::is_fixed_point<TypeLhs>() and
+                             std::is_same_v<TypeLhs, TypeRhs>>* = nullptr>
```
[the remark this answers is not in the document]
Comment (codereport) [reply cpp/src/binaryop/compiled/operation.cuh:169]: It could. In general though we don't support binaryops across different decimal types.

File: cpp/src/binaryop/binaryop.cpp (hunk 2 above)
Comment (codereport) [reply cpp/src/binaryop/binaryop.cpp:115]: I can add it.

File: cpp/include/cudf/fixed_point/fixed_point.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (bdice) [reply cpp/include/cudf/fixed_point/fixed_point.hpp:454]: Ah, I see now. Resolving.

**Review by bdice — COMMENTED** [review]
A few comments on `is_modulo_binop`.

File: cpp/src/binaryop/binaryop.cpp [hunk 6]
```diff
@@ -105,12 +105,22 @@ bool is_comparison_binop(binary_operator op)
...
 }
 
+/**
+ * @brief Returns `true` if `binary_operator` `op` is a modulus operation
+ */
+bool is_modulo_binop(binary_operator op)
+{
+  return op == binary_operator::MOD or   ///< operator %
+         op == binary_operator::PMOD or  ///< positive modulo operator
+         op == binary_operator::PYMOD;   ///< operator % but following python's negative sign rules
```
[the remark this answers is not in the document]
Comment (codereport) [reply cpp/src/binaryop/binaryop.cpp:115]: Yea, I can switch them to `//`, I was just coping from the scoped Enum class

File: cpp/src/binaryop/binaryop.cpp [hunk 7]
```diff
@@ -105,12 +105,22 @@ bool is_comparison_binop(binary_operator op)
          op == binary_operator::NULL_EQUALS;      // 2 null = true; 1 null = false; else ==
 }
 
+/**
+ * @brief Returns `true` if `binary_operator` `op` is a modulus operation
+ */
+bool is_modulo_binop(binary_operator op)
```
[the remark this answers is not in the document]
Comment (codereport) [reply cpp/src/binaryop/binaryop.cpp:111]: Yea, I took a look and agree. I liked @robertmaynard's suggestion as well, but including them with the "basic" ones if the two `NULL_MIN/MAX` are included is my preference as well. Will update.

File: cpp/src/binaryop/compiled/operation.cuh [hunk 8]
```diff
@@ -162,12 +162,24 @@ struct PMod {
...
+  __device__ inline auto operator()(TypeLhs x, TypeRhs y)
```
Comment (bdice) [inline cpp/src/binaryop/compiled/operation.cuh:170]: I noticed that every struct in this file has trailing return types except `PMod`. Might be worth addressing in a separate PR.

**Review by bdice — APPROVED** [review]
LGTM!

File: cpp/src/binaryop/compiled/operation.cuh (hunk 8 above)
Comment (codereport) [reply cpp/src/binaryop/compiled/operation.cuh:170]: Probably better to save for another PR.

**Comment by codereport** [comment]
@gpucibot merge
